### PR TITLE
[KYLO-1722] - Refactor and fix RefreshableDataSource

### PR DESCRIPTION
**Pull request by binhnv**
Hi,

This is the change to refactor `RefreshableDataSource` class.
This change also fixes the `Invalid session handler` reported in [KYLO-1722](https://kylo-io.atlassian.net/browse/KYLO-1722) since any returned connection is validated.

-Binh

### CODE REVIEW

File: integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-hadoop-service-bundle/nifi-hadoop-service/src/main/java/com/thinkbiganalytics/nifi/v2/thrift/RefreshableDataSource.java [hunk 1]
```diff
@@ -75,274 +43,60 @@
...
-     * called to refresh the connection if needed
-     */
-    private void refresh() {
-        if (isRefreshing.compareAndSet(false, true)) {
-            log.info("REFRESHING DATASOURCE for {} ", this.url);
-            target.set(create());
-            isRefreshing.set(false);
-        } else {
-            //unable to refresh.  Refresh already in progress
+    public RefreshableDataSource(String driverClassName,
```
Comment (KyloTech) [inline integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-hadoop-service-bundle/nifi-hadoop-service/src/main/java/com/thinkbiganalytics/nifi/v2/thrift/RefreshableDataSource.java]: MAJOR Constructor has 9 parameters, which is greater than 7 authorized. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00107)

File: integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-hadoop-service-bundle/nifi-hadoop-service/src/main/java/com/thinkbiganalytics/nifi/v2/thrift/HivePoolableConnectionFactory.java [hunk 2]
```diff
@@ -0,0 +1,150 @@
+package com.thinkbiganalytics.nifi.v2.thrift;
+
+/*-
+ * #%L
+ * kylo-nifi-hadoop-service
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (KyloTech) [inline integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-hadoop-service-bundle/nifi-hadoop-service/src/main/java/com/thinkbiganalytics/nifi/v2/thrift/HivePoolableConnectionFactory.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by KyloTech** [comment]
SonarQube analysis reported 10 issues
* ![MAJOR][MAJOR] 7 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 8 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [RefreshableDataSource.java#L9](https://github.com/Teradata/kylo/blob/1de2a71e6dfadf7da1ee093f0b5fe73a9cfe8c10/integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-hadoop-service-bundle/nifi-hadoop-service/src/main/java/com/thinkbiganalytics/nifi/v2/thrift/RefreshableDataSource.java#L9): This block of commented-out lines of code should be removed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
1. ![MAJOR][MAJOR] [ThriftConnectionPool.java#L9](https://github.com/Teradata/kylo/blob/1de2a71e6dfadf7da1ee093f0b5fe73a9cfe8c10/integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-hadoop-service-bundle/nifi-hadoop-service/src/main/java/com/thinkbiganalytics/nifi/v2/thrift/ThriftConnectionPool.java#L9): This block of commented-out lines of code should be removed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACommentedOutCodeLine)
1. ![MAJOR][MAJOR] [ThriftConnectionPool.java#L211](https://github.com/Teradata/kylo/blob/1de2a71e6dfadf7da1ee093f0b5fe73a9cfe8c10/integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-hadoop-service-bundle/nifi-hadoop-service/src/main/java/com/thinkbiganalytics/nifi/v2/thrift/ThriftConnectionPool.java#L211): Make this anonymous inner class a lambda [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1604)
1. ![MAJOR][MAJOR] [ThriftConnectionPool.java#L386](https://github.com/Teradata/kylo/blob/1de2a71e6dfadf7da1ee093f0b5fe73a9cfe8c10/integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-hadoop-service-bundle/nifi-hadoop-service/src/main/java/com/thinkbiganalytics/nifi/v2/thrift/ThriftConnectionPool.java#L386): Rename "principal" which hides the field declared at line 185. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AHiddenFieldCheck)
1. ![MAJOR][MAJOR] [ThriftConnectionPool.java#L403](https://github.com/Teradata/kylo/blob/1de2a71e6dfadf7da1ee093f0b5fe73a9cfe8c10/integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-hadoop-service-bundle/nifi-hadoop-service/src/main/java/com/thinkbiganalytics/nifi/v2/thrift/ThriftConnectionPool.java#L403): Change this instance-reference to a static reference. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2209)
1. ![MINOR][MINOR] [ThriftConnectionPool.java#L67](https://github.com/Teradata/kylo/blob/1de2a71e6dfadf7da1ee093f0b5fe73a9cfe8c10/integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-hadoop-service-bundle/nifi-hadoop-service/src/main/java/com/thinkbiganalytics/nifi/v2/thrift/ThriftConnectionPool.java#L67): Override the "equals" method in this class. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2160)
1. ![MINOR][MINOR] [ThriftConnectionPool.java#L360](https://github.com/Teradata/kylo/blob/1de2a71e6dfadf7da1ee093f0b5fe73a9cfe8c10/integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-hadoop-service-bundle/nifi-hadoop-service/src/main/java/com/thinkbiganalytics/nifi/v2/thrift/ThriftConnectionPool.java#L360): Remove the declaration of thrown exception 'org.apache.nifi.processor.exception.ProcessException' which is a runtime exception. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [ThriftConnectionPool.java#L363](https://github.com/Teradata/kylo/blob/1de2a71e6dfadf7da1ee093f0b5fe73a9cfe8c10/integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-hadoop-service-bundle/nifi-hadoop-service/src/main/java/com/thinkbiganalytics/nifi/v2/thrift/ThriftConnectionPool.java#L363): Immediately return this expression instead of assigning it to the temporary variable "con". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1488)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-hadoop-service-bundle/nifi-hadoop-service/src/main/java/com/thinkbiganalytics/nifi/v2/thrift/RefreshableDataSource.java (hunk 1 above)
Comment (KyloTech) [inline integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-hadoop-service-bundle/nifi-hadoop-service/src/main/java/com/thinkbiganalytics/nifi/v2/thrift/RefreshableDataSource.java]: MAJOR Constructor has 9 parameters, which is greater than 7 authorized. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00107)

File: integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-hadoop-service-bundle/nifi-hadoop-service/src/main/java/com/thinkbiganalytics/nifi/v2/thrift/HivePoolableConnectionFactory.java (hunk 2 above)
Comment (KyloTech) [inline integrations/nifi/nifi-nar-bundles/nifi-standard-services/nifi-hadoop-service-bundle/nifi-hadoop-service/src/main/java/com/thinkbiganalytics/nifi/v2/thrift/HivePoolableConnectionFactory.java]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

**Comment by freastro** [comment]
@binhnv, how does this fix KYLO-1722? It looks you rearranged some of the code but it's doing the same thing.

**Comment by binhnv** [comment]
@freastro it fixes the issue because it does the validation on the connection before it returns connection to the caller (If you trace `BasicDataSource` source code you will see that it calls `PoolableConnectionFactory`'s method for validation). The issue with old code is that it validates a connection but the validated connection is not returned to the caller. Instead, it returns a connection from the pool which is not validated.
Also, I refactor the code to simplifies `RefreshableDataSource`.

-Binh

**Pull request closed without merging**